components/selectors/parser.rs

```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
         // NB: A parse-order iterator is a Rev<>, which doesn't expose as_slice(),
         // which we need for |split|. So we split by combinators on a match-order
         // sequence and then reverse.
-        let mut combinators = self.iter_raw_match_order().rev().filter(|x| x.is_combinator());
+
+        let mut combinators = self.iter_raw_match_order().rev().filter(|x| x.is_combinator()).peekable();
         let compound_selectors = self.iter_raw_match_order().as_slice().split(|x| x.is_combinator()).rev();
 
         let mut combinators_exhausted = false;
         for compound in compound_selectors {
             debug_assert!(!combinators_exhausted);
-            for item in compound.iter() {
-                item.to_css(dest)?;
+
+            // https://www.w3.org/TR/cssom-1/#serializing-selectors
+
+            // 1. If there is only one simple selector in the compound selectors
+            //    which is a universal selector, append the result of
+            //    serializing the universal selector to s.
+            //
+            // Check if `compound.len() > 0` first--this can happen if we have
+            // something like `... > ::before`, because we store `>` and `::`
+            // both as combinators internally.
+            //
+            // If we are in this case, we continue to the next iteration of the
+            // `for compound in compound_selectors` loop.
+            let namespace = if compound.len() > 0 {
+                let (namespace, first_non_namespace) = match &compound[0] {
```

Comment:
nit: Why the reference? I think it'd be clearer to do:

```rust
let has_namespace = matches!(compound[0], Component::ExplicitAnyNamespace | ...);
```

And derive the rest from that.